### PR TITLE
Mapping configuration for assessment

### PR DESCRIPTION
**Pull request by Adam-W1**
Map answered questions to assessment sub criteria and sections

https://digital.dclg.gov.uk/jira/browse/FS-1829

structure of config:
```
- cofr2w2 contains (ordered):
 - unscored sections contains [{sub criteria}] contains [{themes}] contains [{answers}]
 - scored criteria contains [{sub criteria}] contains [{themes}] contains [{answers}]
```
An answer dictionary contains 
```
{
    "field_id": ", // used to extract an answer from the matching field_id in the application payload
    "form_name": ", // used to help traverse the applciation payload (unless flattened in which case this can be removed)
    "field_type": "",  // field type at source (in the form json)
    "presentation_type": "", // field type used by assessment service to determine template component useage
    "question": "", // used to provide context - will this be used instead of the question provided on the applciation json?
},
```

### CODE REVIEW

**Review by SentientHousePlant — COMMENTED** [review]
I like this! Makes sense that themes are children of sub criteria. 

If you want to be really extra you could write a schema validator à la https://github.com/keleshev/schema or https://github.com/python-jsonschema/jsonschema so that future schemas conform to a validatable standard.

**Comment by Adam-W1** [comment]
> Some additional food for thought : you could define some structures outside of the hierarchy and use an internal reference to them. This is what we do in the open api spec with components.
> 
> This would allow for some modularity, which might be useful as this structure expands.
> 
> See https://pypi.org/project/jsonref/ for an example of this.

love those suggestions ❤️

File: mapping.json [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        "name": "Organisation information",
+        "sub_criteria": [{
+            "id": "org_info",
+            "name": "Organisation information",
+            "themes": [{
+                "id": "general_info",
+                "name": "General information",
+                "answers": [{
+                    "field_id": "pipopl",
+                    "field_type": "textarea"
```
Comment (gidsg) [inline mapping.json:14]: Could this be derived from the `type` field in the submitted JSON? So maybe we don't need it here as well?

File: mapping.json [hunk 2]
```diff
@@ -0,0 +1,111 @@
+{
+    "schema_id": "cof_r2_assessment",
```
Comment (srh-sloan) [inline mapping.json:2]: should we call it `cof_r2w2_assessment` as it maps directly then to the short names of fund/round

File: mapping.json (hunk 1 above)
Comment (Adam-W1) [reply mapping.json:14]: I had this discussion with @stevewalton28 and @harryyo yesterday. 
The reason we kept it in is to have a simple toggle when deciding how to render the field, the runner might say it's an XX type, but we may want to display it in a text field template etc. If I have understood correctly, It gives more responsibility to the assessment side in presentation

File: mapping.json [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        "id": "org_info",
+        "name": "Organisation information",
+        "sub_criteria": [{
+            "id": "org_info",
+            "name": "Organisation information",
+            "themes": [{
+                "id": "general_info",
+                "name": "General information",
+                "answers": [{
+                    "field_id": "pipopl",
```
[the remark this answers is not in the document]
Comment (Adam-W1) [reply mapping.json:13]: Nice suggestion 💯

File: mapping.json (hunk 2 above)
Comment (Adam-W1) [reply mapping.json:2]: sounds sensible :)

File: config/mappings/test_mapping_config.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+with open("config/mappings/mapping.json", "r") as mapping_file:
+    mapping = json.load(mapping_file)
+
+try:
+    validate(instance=mapping, schema=schema)
+except Exception:
+    raise Exception
+finally:
+    print("Mapping conforms with schema.")
```
Comment (SentientHousePlant) [inline config/mappings/test_mapping_config.py:13]: Validate does this for us!

In the future there will be many mapping files. So you could loop through these files using:
```python
for file in os.listdir(directory):
   mapping = json.load(mapping_file)
   validate(instance=mapping, schema=schema)
# Runs if nothing weird happens in for loop.
else:
   print("All mappings are valid!")
```

```suggestion
validate(instance=mapping, schema=schema)
print("Mapping conforms with schema.")
```

Comment (srh-sloan) [inline config/mappings/test_mapping_config.py:13]: Doesn't finally run even if validation fails? So I think you need to move the `print("Mapping conforms with schema.")` inside the `try:` clause

File: mapping.json (hunk 1 above)
Comment (gidsg) [reply mapping.json]: OK makes sense.

File: .flaskenv [hunk 5]
```diff
@@ -2,3 +2,4 @@ FLASK_APP=app.py
 FLASK_ENV=development
 FLASK_RUN_PORT=5000
 FLASK_RUN_HOST=localhost
+DATABASE_URL=postgresql://postgres:postgres@127.0.0.1:5432/assessment_spike
```
Comment (gidsg) [inline .flaskenv:5]: Not sure we want to commit this?

Comment (Adam-W1) [reply .flaskenv:5]: My bad 😱

**Review by srh-sloan — APPROVED** [review]
Looks good, I like that it's split into a few smaller files now too 👍

File: config/mappings/test_mapping_config.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import json
+from jsonschema import validate
+from assessment_mapping_schema import top_level_assessment_mapping_schema as schema
+from assessment_mapping import cof_r2w2_assessment as mapping
+
+validate(instance=mapping, schema=schema)
+print("Mapping conforms with schema.")
```
Comment (tferns) [inline config/mappings/test_mapping_config.py:7]: This could maybe be added to the tests that are ran by `pytest`?

Comment (Adam-W1) [reply config/mappings/test_mapping_config.py:7]: Good call, ill hold off for now as this might be moving to assessment

**Pull request merged**